source/s3_meta_request.c

```diff
@@ -1385,14 +1385,16 @@ static void s_s3_meta_request_stream_complete(struct aws_http_stream *stream, in
     if (meta_request->checksum_config.validate_response_checksum) {
         s_get_response_part_finish_checksum_helper(connection, error_code);
     }
-    if (error_code != AWS_ERROR_S3_CANCELED && error_code != AWS_ERROR_S3_PAUSED) {
-        /* BEGIN CRITICAL SECTION */
+    /* BEGIN CRITICAL SECTION */
+    {
         aws_s3_meta_request_lock_synced_data(meta_request);
-        AWS_ASSERT(request->synced_data.http_stream != NULL);
-        aws_linked_list_remove(&request->ongoing_http_requests_list_node);
+        if (request->synced_data.http_stream) {
+            /* The request has been cancelled, and the node has been removed. */
+            aws_linked_list_remove(&request->ongoing_http_requests_list_node);
```

Comment:
I didn't review the [PR](https://github.com/awslabs/aws-c-s3/pull/395) where this was added, but I think there are some more race conditions...

Reply:
https://github.com/awslabs/aws-c-s3/blob/c91f4c3c3bbee8dcd4a49cff0713825aef33e7cb/source/s3_meta_request.c#L1074-L1094

We activate the stream, then acquire the lock, then modify `http_stream` and `ongoing_http_requests_list`.

Once you call activate(), the stream may complete before we acquire the lock. So the stream would try to take itself out of the list. Then we'd put the stream in the list. This probably leads to a leak or dangling reference.

Reply:
Another race condition:
- The meta-request is cancelled, so we iterate the `ongoing_http_requests_list` and cancel streams
- Then on another thread we just finished activating a stream and put it into the list, it runs to completion instead of getting canceled.

So from this function we probably want to check whether we should still be making the request, and NOT make the request, or not activate it (is that allowed?), or cancel it instead of adding to list

Reply:
also: does `http_stream` solely exist so we can cancel it? If so, let's make that a bit more clear by doing things like:
- set it NULL after removing from `ongoing_http_requests`
- rename it ~http_stream~ -> `ongoing_http_stream` to make the connection to `ongoing_http_requests` more clear